dask/dataframe/multi.py

```diff
@@ -921,8 +926,20 @@ def stack_partitions(dfs, divisions, join="outer"):
     for df in dfs:
         # dtypes of all dfs need to be coherent
         # refer to https://github.com/dask/dask/issues/4685
+        # and https://github.com/dask/dask/issues/5968.
+        if is_dataframe_like(df):
+            copied = False
```

Comment:
Rather than this, how about a check for any mismatching types.

```python
current_dtypes = df.dtypes
expected_dtypes = meta.dtypes

needs_astype = (current_dtypes != expected_dtypes ) & df.dtypes.apply(is_categorical_dtype)


if len(needs_astype):
    df = df.copy()
    df[needs_astype] = df[needs_astype].astype(**expected_dtypes.todict())

```

Reply:
That'll also give a bit smaller task graph, since we make a single call to astype.